<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  
  <title>在R中绘制进化树：ggtree学习笔记 | Jinlong Zhang&#39;s blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="ggtree （https:&#x2F;&#x2F;guangchuangyu.github.io&#x2F;software&#x2F;ggtree&#x2F;）是香港大学余光创博士编写的R程序包。ggtree是ggplot2程序包的扩展 (http:&#x2F;&#x2F;www.ggplot2-exts.org&#x2F;)，能够像ggplot2一样，用图层化的语法绘制进化树。这与APE用参数控制绘图明显不同。与ggplot2一样，在ggtree中，通过不同的图层组合即">
<meta property="og:type" content="article">
<meta property="og:title" content="在R中绘制进化树：ggtree学习笔记">
<meta property="og:url" content="https://helixcn.github.io/2016/04/10/2016-4-10-ggtree/index.html">
<meta property="og:site_name" content="Jinlong Zhang&#39;s blog">
<meta property="og:description" content="ggtree （https:&#x2F;&#x2F;guangchuangyu.github.io&#x2F;software&#x2F;ggtree&#x2F;）是香港大学余光创博士编写的R程序包。ggtree是ggplot2程序包的扩展 (http:&#x2F;&#x2F;www.ggplot2-exts.org&#x2F;)，能够像ggplot2一样，用图层化的语法绘制进化树。这与APE用参数控制绘图明显不同。与ggplot2一样，在ggtree中，通过不同的图层组合即">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2016-04-10T07:21:54.000Z">
<meta property="article:modified_time" content="2021-03-26T05:34:24.572Z">
<meta property="article:author" content="Jinlong Zhang">
<meta property="article:tag" content="可视化">
<meta name="twitter:card" content="summary">
  
    <link rel="alternate" href="/atom.xml" title="Jinlong Zhang's blog" type="application/atom+xml">
  
  
    <link rel="shortcut icon" href="/favicon.png">
  
  
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-source-code-pro@0.0.71/index.min.css">

  
  
<link rel="stylesheet" href="/css/style.css">

  
    
<link rel="stylesheet" href="/fancybox/jquery.fancybox.min.css">

  
<meta name="generator" content="Hexo 5.4.0"></head>

<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">Jinlong Zhang&#39;s blog</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
          <a class="main-nav-link" href="/about">About</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="Search"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="https://helixcn.github.io"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-2016-4-10-ggtree" class="h-entry article article-type-post" itemprop="blogPost" itemscope itemtype="https://schema.org/BlogPosting">
  <div class="article-meta">
    <a href="/2016/04/10/2016-4-10-ggtree/" class="article-date">
  <time class="dt-published" datetime="2016-04-10T07:21:54.000Z" itemprop="datePublished">2016-04-10</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="p-name article-title" itemprop="headline name">
      在R中绘制进化树：ggtree学习笔记
    </h1>
  

      </header>
    
    <div class="e-content article-entry" itemprop="articleBody">
      
        <p>ggtree （<a target="_blank" rel="noopener" href="https://guangchuangyu.github.io/software/ggtree/%EF%BC%89%E6%98%AF%E9%A6%99%E6%B8%AF%E5%A4%A7%E5%AD%A6%E4%BD%99%E5%85%89%E5%88%9B%E5%8D%9A%E5%A3%AB%E7%BC%96%E5%86%99%E7%9A%84R%E7%A8%8B%E5%BA%8F%E5%8C%85%E3%80%82ggtree%E6%98%AFggplot2%E7%A8%8B%E5%BA%8F%E5%8C%85%E7%9A%84%E6%89%A9%E5%B1%95">https://guangchuangyu.github.io/software/ggtree/）是香港大学余光创博士编写的R程序包。ggtree是ggplot2程序包的扩展</a> (<a target="_blank" rel="noopener" href="http://www.ggplot2-exts.org/)%EF%BC%8C%E8%83%BD%E5%A4%9F%E5%83%8Fggplot2%E4%B8%80%E6%A0%B7%EF%BC%8C%E7%94%A8%E5%9B%BE%E5%B1%82%E5%8C%96%E7%9A%84%E8%AF%AD%E6%B3%95%E7%BB%98%E5%88%B6%E8%BF%9B%E5%8C%96%E6%A0%91%E3%80%82%E8%BF%99%E4%B8%8EAPE%E7%94%A8%E5%8F%82%E6%95%B0%E6%8E%A7%E5%88%B6%E7%BB%98%E5%9B%BE%E6%98%8E%E6%98%BE%E4%B8%8D%E5%90%8C%E3%80%82%E4%B8%8Eggplot2%E4%B8%80%E6%A0%B7%EF%BC%8C%E5%9C%A8ggtree%E4%B8%AD%EF%BC%8C%E9%80%9A%E8%BF%87%E4%B8%8D%E5%90%8C%E7%9A%84%E5%9B%BE%E5%B1%82%E7%BB%84%E5%90%88%E5%8D%B3%E5%8F%AF%E7%BB%98%E5%88%B6%E5%87%BA%E6%9B%B4%E4%B8%BA%E5%A4%8D%E6%9D%82%E7%9A%84%E5%9B%BE%E5%BD%A2%E3%80%82">http://www.ggplot2-exts.org/)，能够像ggplot2一样，用图层化的语法绘制进化树。这与APE用参数控制绘图明显不同。与ggplot2一样，在ggtree中，通过不同的图层组合即可绘制出更为复杂的图形。</a></p>
<p>该程序包在正式发布之前就受到CRAN Task Views:Phylogenetics 的作者Brian O’Meara 的关注。ggtree发布在Bioconductor之后，更是受到国际同行的广泛赞誉。作者介绍ggtree的论文在英国生态学会的Methods in Ecology and Evolution发表之后， 在不到一年的时间里， 引用次数已经超过100，可见受欢迎程度。</p>
<p>在数据分析中，读取进化树一般是用ape程序包的read.tree(), 但是构建进化树时，不同软件生成的文件格式多种多样，虽然大部分都是以newick或者nexus格式为基础，但是有时要将进化树建立过程中的信息标注在进化树上，这时read.tree读取的信息往往不足，编写函数导入相关信息也比较麻烦。鉴于此，作者为ggtree编写了读取进化树的函数，以弥补ape程序包read.tree的不足。ggtree能够读取并转换BEAST, r8s，RAxML等软件所生成的数据。用户也可自定义数据，轻松实现R实现基于进化树的高级绘图。</p>
<p>余博士关于ggtree的介绍：</p>
<p><a target="_blank" rel="noopener" href="http://cos.name/2015/11/to-achieve-the-visualization-and-annotation-of-evolutionary-tree-using-ggtree/">http://cos.name/2015/11/to-achieve-the-visualization-and-annotation-of-evolutionary-tree-using-ggtree/</a></p>
<p>本笔记主要基于：<br><a target="_blank" rel="noopener" href="http://www.bioconductor.org/packages/release/bioc/vignettes/ggtree/inst/doc/ggtree.html">http://www.bioconductor.org/packages/release/bioc/vignettes/ggtree/inst/doc/ggtree.html</a></p>
<pre><code class="R">#### ggtree可以直接读取的数据格式

Newick (via ape)
Nexus (via ape)
New Hampshire eXtended format (NHX) （http://home.cc.umanitoba.ca/~psgendb/doc/atv/NHX.pdf）
Jplace （http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0031009）
Phylip

以及以下软件输出的数据:

BEAST2
EPA3
HYPHY5
PAML1
PHYLDOG6
pplacer4
r8s7
RAxML8
RevBayes9

####################################################
用以下函数读取
read.beast()       ## for parsing output of BEASE
read.codeml()      ## for parsing output of CODEML (rst and mlc files)
read.codeml_mlc()  ## for parsing mlc file (output of CODEML)
read.hyphy()       ## for parsing output of HYPHY
read.jplace()      ## for parsing jplace file including output from EPA and pplacer
read.nhx()         ## for parsing NHX file including output from PHYLODOG and RevBayes
read.paml_rst()    ## for parsing rst file (output of BASEML and CODEML)
read.r8s()         ## for parsing output of r8s
read.raxml()       ## for parsing output of RAxML

#####################################################
###### 定义的 S4 Classes #####################################
apeBootstrap ##for bootstrap analysis of ape::boot.phylo()10, output of apeBoot() defined in ggtree
beast        ##for storing output of read.beast()
codeml       ## for storing output of read.codeml()
codeml_mlc   ##for storing output of read.codeml_mlc()
hyphy        ## for storing output of read.hyphy()
jplace       ## for storing output of read.jplace()
nhx          ## for storing output of read.nhx()
paml_rst     ## for rst file obtained by PAML, including BASEML and CODEML.
phangorn     ## for storing ancestral sequences inferred by R package phangorn11, output of phyPML defined in ggtree
r8s          ## for storing output of read.r8s()
raxml        ## for storing output of read.raxml()

### 其他支持的S4 类
###

phylo, 
multiPhylo (defined by ape10), 
phylo4 (defined by phylobase) 
obkData (defined in OutbreakTools) and 
phyloseq (defined in phyloseq).

### 获取任何能够在进化树上显示的参数
get.fields

### 显示进化树
ggtree(tree_object)

### 将任何格式的进化树转换为 phylo格式
get.tree

### 界定分类单元
groupOTU()
groupClade()

### 将进化树转换为 data.frame
fortify

############################################
#########3 举例： BEAST

library(ggtree)
file &lt;- system.file(&quot;extdata/BEAST&quot;, &quot;beast_mcc.tree&quot;, package=&quot;ggtree&quot;)
beast &lt;- read.beast(file)
beast

get.fields(beast)
ggtree(beast, ndigits = 2, branch.length = &#39;none&#39;) + geom_text(aes(x = branch, label =length_0.95_HPD), vjust = -.5, color = &#39;firebrick&#39;)

######### 导入的数据通过 fortify函数能够转换为 data.frame
beast_data &lt;- fortify(beast)
head(beast_data)


################ PAML
brstfile &lt;- system.file(&quot;extdata/PAML_Baseml&quot;, &quot;rst&quot;, package=&quot;ggtree&quot;)
brst &lt;- read.paml_rst(brstfile)
brst

p &lt;- ggtree(brst) +
   geom_text(aes(x = branch, label = marginal_AA_subs), vjust=-.5, color=&#39;steelblue&#39;) +
   theme_tree2()
print(p)

################# CODEML
crstfile &lt;- system.file(&quot;extdata/PAML_Codeml&quot;, &quot;rst&quot;, package=&quot;ggtree&quot;)
crst &lt;- read.paml_rst(crstfile)
crst

##### %&lt;% 类似于格式刷， 将某一进化树绘制时的参数， 转移到要绘制的进化树中
p %&lt;% crst

################ 
#### CODEML mlc文件
mlcfile &lt;- system.file(&quot;extdata/PAML_Codeml&quot;, &quot;mlc&quot;, package=&quot;ggtree&quot;)
mlc &lt;- read.codeml_mlc(mlcfile)
mlc

ggtree(mlc) + geom_text(aes(x=branch, label=dN_vs_dS), color=&#39;blue&#39;, vjust=-.2)

#### 获取有用的信息
get.fields(mlc)

#### 指定需要调整的枝长

ggtree(mlc, branch.length = &quot;dN_vs_dS&quot;, aes(color=dN_vs_dS)) +
   scale_color_continuous(name=&#39;dN/dS&#39;, limits=c(0, 1.5),
                          oob=scales::squish, low=&quot;darkgreen&quot;, high=&quot;red&quot;)+
   theme_tree2(legend.position=c(.9, .5))

#### CODEML 的 rst以及 mlc文件

ml &lt;- read.codeml(crstfile, mlcfile)
ml
#### 
head(fortify(ml))

#### 读取 HYPHY文件并绘制进化树

nwk &lt;- system.file(&quot;extdata/HYPHY&quot;, &quot;labelledtree.tree&quot;, package=&quot;ggtree&quot;)
ancseq &lt;- system.file(&quot;extdata/HYPHY&quot;, &quot;ancseq.nex&quot;, package=&quot;ggtree&quot;)
tipfas &lt;- system.file(&quot;extdata&quot;, &quot;pa.fas&quot;, package=&quot;ggtree&quot;)
hy &lt;- read.hyphy(nwk, ancseq, tipfas)
hy

ggtree(hy) + geom_text(aes(x=branch, label=AA_subs), vjust=-.5)


##### 读取r8s生成的文件并绘制进化树
r8s &lt;- read.r8s(system.file(&quot;extdata/r8s&quot;, &quot;H3_r8s_output.log&quot;, package=&quot;ggtree&quot;))
ggtree(r8s, branch.length=&quot;TREE&quot;, mrsd=&quot;2014-01-01&quot;) + theme_tree2()

ggtree(get.tree(r8s), aes(color=.id)) + facet_wrap(~.id, scales=&quot;free_x&quot;)

##### 读取RAxML生成的bootstrap文件并绘制进化树

raxml_file &lt;- system.file(&quot;extdata/RAxML&quot;, &quot;RAxML_bipartitionsBranchLabels.H3&quot;,package=&quot;ggtree&quot;)
raxml &lt;- read.raxml(raxml_file)
ggtree(raxml) + geom_label(aes(label=bootstrap, fill=bootstrap)) + geom_tiplab() +
   scale_fill_continuous(low=&#39;darkgreen&#39;, high=&#39;red&#39;) + theme_tree2(legend.position=&#39;right&#39;)

##### 读取NHX New Hampshire eXtended文件并绘制进化树
nhxfile &lt;- system.file(&quot;extdata&quot;, &quot;ADH.nhx&quot;, package=&quot;ggtree&quot;)
nhx &lt;- read.nhx(nhxfile)
ggtree(nhx) + geom_tiplab() + geom_point(aes(color=S), size=5, alpha=.5) + 
   theme(legend.position=&quot;right&quot;) +
   geom_text(aes(label=branch.length, x=branch), vjust=-.5) + 
   xlim(NA, 0.3)

##### 读取PHYLIP文件并绘制进化树以及序列比对图
phyfile &lt;- system.file(&quot;extdata&quot;, &quot;sample.phy&quot;, package=&quot;ggtree&quot;)
phylip &lt;- read.phylip(phyfile)
phylip

ggtree(phylip) + geom_tiplab()

msaplot(phylip, offset=1)

##### jplace格式数据同时保存自定义数据

jpf &lt;- system.file(&quot;extdata/sample.jplace&quot;,  package=&quot;ggtree&quot;)
jp &lt;- read.jplace(jpf)
print(jp)
## get only best hit
get.placements(jp, by=&quot;best&quot;)

## get all placement
get.placements(jp, by=&quot;all&quot;)

########################################################################
############ 进化树合并 ### 分类单元匹配 ##################################
############ 作者讲的不太清楚 
beast_file &lt;- system.file(&quot;examples/MCC_FluA_H3.tree&quot;, package=&quot;ggtree&quot;)
beast_tree &lt;- read.beast(beast_file)

rst_file &lt;- system.file(&quot;examples/rst&quot;, package=&quot;ggtree&quot;)
mlc_file &lt;- system.file(&quot;examples/mlc&quot;, package=&quot;ggtree&quot;)
codeml_tree &lt;- read.codeml(rst_file, mlc_file)

merged_tree &lt;- merge_tree(beast_tree, codeml_tree)

merged_tree

#################################################################
#### 读取 jplace 文件， 并将保存在csv文件种的性状信息

tree &lt;- system.file(&quot;extdata&quot;, &quot;pa.nwk&quot;, package=&quot;ggtree&quot;)
data &lt;- read.csv(system.file(&quot;extdata&quot;, &quot;pa_subs.csv&quot;, package=&quot;ggtree&quot;),stringsAsFactor=FALSE)
print(tree)

outfile &lt;- tempfile()
write.jplace(tree, data, outfile)
jp &lt;- read.jplace(outfile)

ggtree(jp) + 
   geom_text(aes(x=branch, label=subs), color=&quot;purple&quot;, vjust=-1, size=3) + 
   geom_text(aes(label=gc), color=&quot;steelblue&quot;, hjust=-.6, size=3) +
   geom_tiplab()

#############################################################################
########## 绘制进化树
library(&quot;ggtree&quot;)
nwk &lt;- system.file(&quot;extdata&quot;, &quot;sample.nwk&quot;, package=&quot;ggtree&quot;)
tree &lt;- read.tree(nwk)

ggplot(tree, aes(x, y)) + geom_tree() + theme_tree()
ggtree(tree, color=&quot;firebrick&quot;, size=1, linetype=&quot;dotted&quot;)

### ggtree会自动将进化树转换为 ladderized， 如果要取消 ladderized， 则应该改为 ladderize
ggtree(tree, ladderize=FALSE)

### 不要枝长， 只查
ggtree(tree, branch.length=&quot;none&quot;)

########## 进化树显示的方式 ############

矩形 rectangular (by default)
偏斜 slanted
扇形/圆形 fan or circular

ggtree(tree) + ggtitle(&quot;(Phylogram) rectangular layout&quot;)
ggtree(tree, layout=&quot;slanted&quot;) + ggtitle(&quot;(Phylogram) slanted layout&quot;)
ggtree(tree, layout=&quot;circular&quot;) + ggtitle(&quot;(Phylogram) circular layout&quot;)

###### 只显示拓扑结构
### 矩形
ggtree(tree, branch.length=&#39;none&#39;) + ggtitle(&quot;(Cladogram) rectangular layout&quot;)

### 偏斜
ggtree(tree, layout=&quot;slanted&quot;, branch.length=&#39;none&#39;) + ggtitle(&quot;(Cladogram) slanted layout&quot;)

### 圆形
ggtree(tree, layout=&quot;circular&quot;, branch.length=&quot;none&quot;) + ggtitle(&quot;(Cladogram) circular layout&quot;)

##### 无根树
ggtree(tree, layout=&quot;unrooted&quot;) + ggtitle(&quot;unrooted layout&quot;)

#######################################################################################
########### Ultrametric Tree 绘图会自动添加标尺 ###################
ggtree(tree) + geom_treescale()

#### 更改标尺的长短颜色， 粗细 
ggtree(tree)+geom_treescale(x=0, y=12, width=6, color=&#39;red&#39;)

#################### 
ggtree(tree)+geom_treescale(fontsize=8, linesize=2, offset=-1)

###### 添加时间轴 依靠 theme_tree2()
ggtree(tree) + theme_tree2()

#########################################################################################
###### 节点以及分类单元的标识， 根据是否为 tiplabel以及node分别显示
ggtree(tree)+geom_point(aes(shape=isTip, color=isTip), size=3)

#### Node Point参数 以及 tippoint分别显示
p &lt;- ggtree(tree) + geom_nodepoint(color=&quot;#b5e521&quot;, alpha=1/4, size=10)
p + geom_tippoint(color=&quot;#FDAC4F&quot;, shape=8, size=3)

#### Tiplabel的显示
p + geom_tiplab(size=3, color=&quot;purple&quot;)

#### 圆形进化树的标签
ggtree(tree, layout=&quot;circular&quot;) + geom_tiplab(aes(angle=angle), color=&#39;blue&#39;)

#### 进化树tip label显示到branch上
p + geom_tiplab(aes(x=branch), size=3, color=&quot;purple&quot;, vjust=-0.3)

#### %&lt;% 若之前已经有ggtree object完成了进化树的绘制， 那么可以用 %&gt;% 将前一个进化树的属性转移到当前进化树中。 %&lt;% 的作用相当于word中的格式刷。 
p %&lt;% rtree(50)

#################################################################
######### ggtree中有两个主题 theme， theme_tree2() 和 theme_tree()
### theme_tree2 与 后者的区别在于 增加了时间坐标轴。 
### 无坐标轴
ggtree(rtree(30), color=&quot;red&quot;) + theme_tree()

### 有坐标轴
ggtree(rtree(30), color=&quot;red&quot;) + theme_tree2()

### 改变背景颜色
ggtree(rtree(30), color=&quot;red&quot;) + theme_tree(&quot;steelblue&quot;)

### 
ggtree(rtree(20), color=&quot;white&quot;) + theme_tree(&quot;#b5e521&quot;)

#################################################################
############# 显示多个进化树 ######################################
trees &lt;- lapply(c(10, 20, 40), rtree)
class(trees) &lt;- &quot;multiPhylo&quot;
ggtree(trees) + facet_wrap(~.id, scale=&quot;free&quot;) + geom_tiplab()

#################################################################
btrees &lt;- read.tree(system.file(&quot;extdata/RAxML&quot;, &quot;RAxML_bootstrap.H3&quot;, package=&quot;ggtree&quot;))
ggtree(btrees) + facet_wrap(~.id, ncol=10)


#################################################################
########### 显示bootstrap进化树作为背景 #################################
p &lt;- ggtree(btrees, layout=&quot;rectangular&quot;,   color=&quot;lightblue&quot;, alpha=.3)

best_tree &lt;- read.tree(system.file(&quot;extdata/RAxML&quot;, &quot;RAxML_bipartitionsBranchLabels.H3&quot;,package=&quot;ggtree&quot;))
df &lt;- fortify(best_tree, branch.length = &#39;none&#39;)
p + geom_tree(data=df, color=&#39;firebrick&#39;)

################################################################################
############# 进化树的操作 ##########################################

#### 对进化树操作之前， 必须要定位进化树内部的节点， 每一个节点都有一个顺序号
#### ggtree提供了显示内部节点号的函数

nwk &lt;- system.file(&quot;extdata&quot;, &quot;sample.nwk&quot;, package=&quot;ggtree&quot;)
tree &lt;- read.tree(nwk)
ggtree(tree) + geom_text2(aes(subset=!isTip, label=node), hjust=-.3) + geom_tiplab()

#### 用MRCA显示节点编号
MRCA(tree, tip=c(&#39;A&#39;, &#39;E&#39;))

########### 显示某一个节点下的整个类群 clade
p &lt;- ggtree(tree)
viewClade(p + geom_tiplab(), node=21)

########### 基于内部节点号， 划定clade， 按照clade划分颜色
tree &lt;- groupClade(tree, node=21)
ggtree(tree, aes(color=group, linetype=group))

##########  不同group分不同颜色显示， 显示tiplabel时， 不同的组别用不同的颜色表示
tree &lt;- groupClade(tree, node=c(21, 17))
ggtree(tree, aes(color=group, linetype=group)) + geom_tiplab(aes(subset=(group==2)))

########### 手工指定类群的分组，并且表示为不同的颜色 
cls &lt;- list(c1=c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;, &quot;D&quot;, &quot;E&quot;),
           c2=c(&quot;F&quot;, &quot;G&quot;, &quot;H&quot;),
           c3=c(&quot;L&quot;, &quot;K&quot;, &quot;I&quot;, &quot;J&quot;),
           c4=&quot;M&quot;)

#### s输入对象未 ggtree时
tree &lt;- groupOTU(tree, cls)
library(&quot;colorspace&quot;)
ggtree(tree, aes(color=group, linetype=group)) + geom_tiplab() +
    scale_color_manual(values=c(&quot;black&quot;, rainbow_hcl(4))) + theme(legend.position=&quot;right&quot;)

#### 输入对象为
p &lt;- ggtree(tree)
groupOTU(p, LETTERS[1:5]) + aes(color=group) + geom_tiplab() +scale_color_manual(values=c(&quot;black&quot;, &quot;firebrick&quot;))


##### groupOTU划分组
library(&quot;ape&quot;)
data(chiroptera)
groupInfo &lt;- split(chiroptera$tip.label, gsub(&quot;_\w+&quot;, &quot;&quot;, chiroptera$tip.label))
chiroptera &lt;- groupOTU(chiroptera, groupInfo)
ggtree(chiroptera, aes(color=group), layout=&#39;circular&#39;) + geom_tiplab(size=1,aes(angle=angle))


##### 类群折叠与展开
cp &lt;- ggtree(tree) %&gt;% collapse(node=21)

### 折叠节点号为21的类群
cp + geom_point2(aes(subset=(node == 21)), size=5, shape=23, fill=&quot;steelblue&quot;)

### 展开
cp %&gt;% expand(node=21)

### 折叠以及展开多个节点
p1 &lt;- ggtree(tree)
p2 &lt;- collapse(p1, 21) + geom_point2(aes(subset=(node==21)), size=5, shape=23, fill=&quot;blue&quot;)
p3 &lt;- collapse(p2, 17) + geom_point2(aes(subset=(node==17)), size=5, shape=23, fill=&quot;red&quot;)
p4 &lt;- expand(p3, 17)
p5 &lt;- expand(p4, 21)

multiplot(p1, p2, p3, p4, p5, ncol=5)


#### 类群的缩放
multiplot(ggtree(tree) + geom_hilight(21, &quot;steelblue&quot;),
         ggtree(tree) %&gt;% scaleClade(21, scale=0.3) + geom_hilight(21, &quot;steelblue&quot;),
         ncol=2)

#### 指定节点编号， 高亮某一个类群
ggtree(tree) + geom_hilight(21, &quot;steelblue&quot;)

#### 类群clade旋转180度 (此例子不能运行)
tree &lt;- groupClade(tree, c(21, 17))
### p &lt;- ggtree(tree, aes(color=group)) + scale_color_manual(values=c(&quot;black&quot;, &quot;firebrick&quot;, &quot;steelblue&quot;))
#### p2 &lt;- rotate(p, 21) %&gt;% rotate(p, 17)
multiplot(p, p2, ncol=2)

########################################################################
##### 两个分类群位置互换
multiplot(p, flip(p, 17, 21), ncol=2)

############################################################################
############################################################################
##### 进化树
beast_file &lt;- system.file(&quot;examples/MCC_FluA_H3.tree&quot;, package=&quot;ggtree&quot;)
beast_tree &lt;- read.beast(beast_file)
beast_tree


##### 绘制进化树
p1 &lt;- ggtree(beast_tree, mrsd=&#39;2013-01-01&#39;) + theme_tree2() +
   ggtitle(&quot;Divergence time&quot;)
p2 &lt;- ggtree(beast_tree, branch.length = &#39;rate&#39;) + theme_tree2() +
   ggtitle(&quot;Substitution rate&quot;)
multiplot(p1, p2, ncol=2)

###### 局部放大某一个分类群 clade
library(&quot;ape&quot;)
data(chiroptera)
library(&quot;ggtree&quot;)
gzoom(chiroptera, grep(&quot;Plecotus&quot;, chiroptera$tip.label))

###### 局部放大某分类群
groupInfo &lt;- split(chiroptera$tip.label, gsub(&quot;_\w+&quot;, &quot;&quot;, chiroptera$tip.label))
chiroptera &lt;- groupOTU(chiroptera, groupInfo)
p &lt;- ggtree(chiroptera, aes(color=group)) + geom_tiplab() + xlim(NA, 23)
gzoom(p, grep(&quot;Plecotus&quot;, chiroptera$tip.label), xmax_adjust=2)

###### 各分支按照条件设色
ggtree(beast_tree, aes(color = rate)) +
   scale_color_continuous(low=&#39;darkgreen&#39;, high=&#39;red&#39;) +
   theme(legend.position=&quot;right&quot;)

##### 标注某一个clade， 在右侧画竖线
set.seed(2015-12-21)
tree = rtree(30)
p &lt;- ggtree(tree) + xlim(NA, 6)

p + geom_cladelabel(node=45, label=&quot;test label&quot;) + 
   geom_cladelabel(node=34, label=&quot;another clade&quot;) 

##### 右侧标注用的竖线可以对齐
p + geom_cladelabel(node=45, label=&quot;test label&quot;, align=TRUE, offset=.5) +
   geom_cladelabel(node=34, label=&quot;another clade&quot;, align=TRUE, offset=.5)

##### 改变标注竖线的颜色
p+geom_cladelabel(node=45, label=&quot;test label&quot;, align=T, color=&#39;red&#39;) +
   geom_cladelabel(node=34, label=&quot;another clade&quot;, align=T, color=&#39;blue&#39;)

##### 旋转标注类群的文字
p+geom_cladelabel(node=45, label=&quot;test label&quot;, align=T, angle=270, hjust=&#39;center&#39;,offset.text=.5) + geom_cladelabel(node=34, label=&quot;another clade&quot;, align=T, angle=45)     

##### 改变标注竖线的粗细， 以及文字的大小
p+geom_cladelabel(node=45, label=&quot;test label&quot;, align=T, angle=270, hjust=&#39;center&#39;,offset.text=.5, barsize=1.5) +
   geom_cladelabel(node=34, label=&quot;another clade&quot;, align=T, angle=45, fontsize=8) 

##### 为文字添加边框， 并设置文本框的背景颜色
p+ geom_cladelabel(node=34, label=&quot;another clade&quot;, align=T, geom=&#39;label&#39;, fill=&#39;lightblue&#39;) 

########################################################################
################# 对类群进行高亮显示 ######################################
nwk &lt;- system.file(&quot;extdata&quot;, &quot;sample.nwk&quot;, package=&quot;ggtree&quot;)
tree &lt;- read.tree(nwk)
ggtree(tree) + geom_hilight(node=21, fill=&quot;steelblue&quot;, alpha=.6) +
   geom_hilight(node=17, fill=&quot;darkgreen&quot;, alpha=.6)

#### 圆形树的类群高亮
ggtree(tree, layout=&quot;circular&quot;) + geom_hilight(node=21, fill=&quot;steelblue&quot;, alpha=.6) +
   geom_hilight(node=23, fill=&quot;darkgreen&quot;, alpha=.6)

#########################################################################
#### 标注 bootstrap数值 
#### 用ape生成一个NJ树， 并且进行bootstrap， 将bootstrap的结果在NJ树的节点上显示。 

library(ape)
data(woodmouse)
d &lt;- dist.dna(woodmouse)
tr &lt;- nj(d)
bp &lt;- boot.phylo(tr, woodmouse, function(xx) nj(dist.dna(xx)))

##### 
tree &lt;- apeBoot(tr, bp)  ### merge phylo and output of boot.phylo to &#39;apeBootstrap&#39; object
ggtree(tree) + geom_label(aes(label=bootstrap)) + geom_tiplab()


##### 对phangorn生成进化树的标注
library(phangorn)
treefile &lt;- system.file(&quot;extdata&quot;, &quot;pa.nwk&quot;, package=&quot;ggtree&quot;)
tre &lt;- read.tree(treefile)
tipseqfile &lt;- system.file(&quot;extdata&quot;, &quot;pa.fas&quot;, package=&quot;ggtree&quot;)
tipseq &lt;- read.phyDat(tipseqfile,format=&quot;fasta&quot;)
fit &lt;- pml(tre, tipseq, k=4)
#### fit是设定条件生成的最优进化树
fit &lt;- optim.pml(fit, optNni=FALSE, optBf=T, optQ=T,
                optInv=T, optGamma=T, optEdge=TRUE,
                optRooted=FALSE, model = &quot;GTR&quot;)

#### 将pml格式的数据转换为 ggtree能够识别的类型
phangorn &lt;- phyPML(fit, type=&quot;ml&quot;)
ggtree(phangorn) + geom_text(aes(x=branch, label=AA_subs, vjust=-.5))

#### ggtree可以读取如下软件生成的进化树， 并且同时保存这些软件输出结果关于进化速率等相应的信息
BEAST
EPA
HYPHY
PAML
PHYLDOG
pplacer
r8s
RAxML
RevBayes

########################################################################
######## 将性状信息与进化树匹配绘图 ###############

nwk &lt;- system.file(&quot;extdata&quot;, &quot;sample.nwk&quot;, package=&quot;ggtree&quot;)
tree &lt;- read.tree(nwk)
p &lt;- ggtree(tree)

##### 生成随机数据
dd &lt;- data.frame(taxa  = LETTERS[1:13], 
                place = c(rep(&quot;GZ&quot;, 5), rep(&quot;HK&quot;, 3), rep(&quot;CZ&quot;, 4), NA),
                value = round(abs(rnorm(13, mean=70, sd=10)), digits=1))
## you don&#39;t need to order the data
## data was reshuffled just for demonstration
dd &lt;- dd[sample(1:13, 13), ]
row.names(dd) &lt;- NULL

##### 数据框匹配字符
##### %&lt;+% 运算符的要求
##### 左侧为进化树
##### 右侧为 data.frame, 要求是该进化树的第一列种的物种名必须能与进化树的tip.label匹配上。 

p &lt;- p %&lt;+% dd + geom_tiplab(aes(color=place)) + 
      geom_tippoint(aes(size = value, shape = place, color = place), alpha = 0.25)
      
p + theme(legend.position=&quot;right&quot;)

##### 将采集地点以及性状值写在枝长上
p + geom_text(aes(color=place, label=place), hjust=1, vjust=-0.4, size=3) +
 geom_text(aes(color=place, label=value), hjust=1, vjust=1.4, size=3)

########################################################################
########################################################################
########################################################################
####################### 在进化树右侧绘制热图 ###############################

beast_file &lt;- system.file(&quot;examples/MCC_FluA_H3.tree&quot;, package=&quot;ggtree&quot;)
beast_tree &lt;- read.beast(beast_file)

genotype_file &lt;- system.file(&quot;examples/Genotype.txt&quot;, package=&quot;ggtree&quot;)
genotype &lt;- read.table(genotype_file, sep=&quot;t&quot;, stringsAsFactor=F)
head(genotype)

p &lt;- ggtree(beast_tree, mrsd=&quot;2013-01-01&quot;) + geom_treescale(x=2008, y=1)
p &lt;- p + geom_tiplab(size=3)
gheatmap(p, genotype, offset = 2, width=0.5)

########################################################################
#### scale_x_ggtree() 保证右侧矩阵的文字说明能正常显示。 
p &lt;- ggtree(beast_tree, mrsd=&quot;2013-01-01&quot;) + geom_tiplab(size=3, align=TRUE) + theme_tree2()
pp &lt;- (p + scale_y_continuous(expand=c(0, 0.3))) %&gt;%
   gheatmap(genotype, offset=4, width=0.5, colnames=FALSE) %&gt;%
       scale_x_ggtree()
pp + theme(legend.position=&quot;right&quot;)

######################## msaplot ###### 
##### 进化树 以及 fasta比对矩阵 
fasta &lt;- system.file(&quot;examples/FluA_H3_AA.fas&quot;, package=&quot;ggtree&quot;)
msaplot(ggtree(beast_tree), fasta) 

######################## 扇形的 msaplot #####
msaplot(ggtree(beast_tree), fasta, window=c(150, 200)) + coord_polar(theta=&#39;y&#39;)

######################## 在进化树的总图上高亮显示某一个clade： geom_hilight
######################## 并且在右侧的大图中放大该clade viewClade subview
set.seed(2016-01-04)
tr &lt;- rtree(30)
tr &lt;- groupClade(tr, node=45)
p &lt;- ggtree(tr, aes(color=group)) + geom_tippoint()
p1 &lt;- p + geom_hilight(node=45)
p2 &lt;- viewClade(p, node=45) + geom_tiplab()
subview(p2, p1+theme_transparent(), x=2.3, y=28.5)


#############################################################################
########## 用inset为内部节点添加直方图， 以显示每个节点在性状上所占比例  ##############
#############################################################################

set.seed(2015-12-31)
tr &lt;- rtree(15)
p &lt;- ggtree(tr)

a &lt;- runif(14, 0, 0.33)
b &lt;- runif(14, 0, 0.33)
c &lt;- runif(14, 0, 0.33)
d &lt;- 1 - a - b - c
dat &lt;- data.frame(a=a, b=b, c=c, d=d)
## input data should have a column of `node` that store the node number 
### 数据必须有一列名为node， 并且值为节点数
dat$node &lt;- 15+1:14

## cols parameter indicate which columns store stats (a, b, c and d in this example)
## cols表示dat的1:4列储存了用于绘图的数据
bars &lt;- nodebar(dat, cols=1:4)

inset(p, bars)

## 柱状图的大小和颜色
inset(p, bars, width=.03, height=.06)

### 柱状图的排列方式， 改为并排排列
bars2 &lt;- nodebar(dat, cols=1:4, position=&#39;dodge&#39;,
                color=c(a=&#39;blue&#39;, b=&#39;red&#39;, c=&#39;green&#39;, d=&#39;cyan&#39;))
p2 &lt;- inset(p, bars2, x=&#39;branch&#39;, width=.03, vjust=-.3)
print(p2)


##############################################################
### 节点上的饼图 

pies &lt;- nodepie(dat, cols=1:4, alpha=.6)
inset(p, pies) 

inset(p, pies, hjust=-.06) ### 饼图向右平移

##############################################################
#### 同时用 pie和bar标注内部节点

pies_and_bars &lt;- bars2
pies_and_bars[9:14] &lt;- pies[9:14]
inset(p, pies_and_bars)

##############################################################
##### 用 inset函数用tip添加对应的箱线图 boxplot， 以展示数据分布

d &lt;- lapply(1:15, rnorm, n=100)
ylim &lt;- range(unlist(d))
bx &lt;- lapply(d, function(y) &#123;
   dd &lt;- data.frame(y=y)
   ggplot(dd, aes(x=1, y=y)) + geom_boxplot() + ylim(ylim) + theme_inset()
&#125;)
names(bx) &lt;- 1:15
inset(p, bx, width=.03, height=.1, hjust=-.05)


#########################################################################
### 混合标注， 未能成功
### p2 &lt;- inset(p, bars2, x=&#39;branch&#39;, width=.03, vjust=-.4)
### p2 &lt;- inset(p2, pies, x=&#39;branch&#39;, vjust=.4)
### bx2 &lt;- lapply(bx, function(g) g+coord_flip())
### inset(p2, bx2, width=.2, height=.03, vjust=.04, hjust=p2datadatax[1:15]-4) + xlim(NA, 4.5)


########################################################################
###### ggplot统计与进化树绘制 ################ 
tr &lt;- rtree(30)
df &lt;- fortify(tr)
df$tipstats &lt;- NA
d1 &lt;- df
d2 &lt;- df
d2$tipstats[d2$isTip] &lt;- abs(rnorm(30))
d1$panel &lt;- &#39;Tree&#39;
d2$panel &lt;- &#39;Stats&#39;
d1$panel &lt;- factor(d1$panel, levels=c(&quot;Tree&quot;, &quot;Stats&quot;))
d2$panel &lt;- factor(d2$panel, levels=c(&quot;Tree&quot;, &quot;Stats&quot;))

p &lt;- ggplot(mapping=aes(x=x, y=y)) + facet_grid(.~panel, scale=&quot;free_x&quot;) + theme_tree2()
p+geom_tree(data=d1) + geom_point(data=d2, aes(x=tipstats)) 

########################################################################
##### 从 phylopic数据库上下载剪影图像， 并标在进化树上

pp &lt;- ggtree(tree) %&gt;% phylopic(&quot;79ad5f09-cf21-4c89-8e7d-0c82a00ce728&quot;, color=&quot;steelblue&quot;,alpha = .3)
print(pp)
</code></pre>
<p>参考：</p>
<p>Yu, G., Smith, D. K., Zhu, H., Guan, Y., &amp; Lam, T. T. Y. (2017). ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. <em>Methods in Ecology and Evolution</em>, <em>8</em>(1), 28-36</p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://helixcn.github.io/2016/04/10/2016-4-10-ggtree/" data-id="ckmq42als00e21k845o7shay6" data-title="在R中绘制进化树：ggtree学习笔记" class="article-share-link">Share</a>
      
      
      
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2016/04/20/2016-4-21-cvlatex/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Newer</strong>
      <div class="article-nav-title">
        
          个人简历的Latex模板
        
      </div>
    </a>
  
  
    <a href="/2016/03/22/2016-3-23-krig/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Older</strong>
      <div class="article-nav-title">克里格空间插值Kriging及其在R的实现</div>
    </a>
  
</nav>

  
</article>


</section>
        
          <aside id="sidebar">
  
    

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tags</h3>
    <div class="widget">
      <ul class="tag-list" itemprop="keywords"><li class="tag-list-item"><a class="tag-list-link" href="/tags/Python/" rel="tag">Python</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/R%E7%BC%96%E7%A8%8B/" rel="tag">R编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E4%BA%BA%E7%89%A9/" rel="tag">人物</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%86%99%E4%BD%9C/" rel="tag">写作</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" rel="tag">可视化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E4%B9%A6/" rel="tag">图书</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" rel="tag">图像处理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%9C%B0%E7%90%86/" rel="tag">地理</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A3%B0%E6%98%8E/" rel="tag">声明</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%A4%A9%E6%96%87/" rel="tag">天文</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" rel="tag">对话录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E5%BF%83%E5%BE%97/" rel="tag">心得</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8A%A5%E5%91%8A/" rel="tag">报告</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%8E%92%E7%89%88/" rel="tag">排版</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%85%E4%B9%A1/" rel="tag">故乡</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" rel="tag">数据库</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" rel="tag">标本馆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" rel="tag">植物分类</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" rel="tag">植物学名</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E6%B1%82%E8%81%8C/" rel="tag">求职</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" rel="tag">版本控制</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81/" rel="tag">生态</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" rel="tag">生态资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" rel="tag">研究资讯</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" rel="tag">系统发育</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BB%9F%E8%AE%A1/" rel="tag">统计</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E7%BC%96%E7%A8%8B/" rel="tag">编程</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" rel="tag">行摄匆匆</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" rel="tag">见闻录</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%AF%97%E6%AD%8C/" rel="tag">诗歌</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BD%AF%E4%BB%B6/" rel="tag">软件</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E8%BF%9B%E5%8C%96/" rel="tag">进化</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E6%83%B3/" rel="tag">随想</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%9A%8F%E7%AC%94/" rel="tag">随笔</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/%E9%A9%B3%E8%AE%BA/" rel="tag">驳论</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Tag Cloud</h3>
    <div class="widget tagcloud">
      <a href="/tags/Python/" style="font-size: 10px;">Python</a> <a href="/tags/R%E7%BC%96%E7%A8%8B/" style="font-size: 18.75px;">R编程</a> <a href="/tags/%E4%BA%BA%E7%89%A9/" style="font-size: 11.88px;">人物</a> <a href="/tags/%E5%86%99%E4%BD%9C/" style="font-size: 13.13px;">写作</a> <a href="/tags/%E5%8F%AF%E8%A7%86%E5%8C%96/" style="font-size: 12.5px;">可视化</a> <a href="/tags/%E5%9B%BE%E4%B9%A6/" style="font-size: 10px;">图书</a> <a href="/tags/%E5%9B%BE%E5%83%8F%E5%A4%84%E7%90%86/" style="font-size: 11.88px;">图像处理</a> <a href="/tags/%E5%9C%B0%E7%90%86/" style="font-size: 17.5px;">地理</a> <a href="/tags/%E5%A3%B0%E6%98%8E/" style="font-size: 10.63px;">声明</a> <a href="/tags/%E5%A4%A9%E6%96%87/" style="font-size: 16.25px;">天文</a> <a href="/tags/%E5%AF%B9%E8%AF%9D%E5%BD%95/" style="font-size: 10px;">对话录</a> <a href="/tags/%E5%BF%83%E5%BE%97/" style="font-size: 10px;">心得</a> <a href="/tags/%E6%8A%A5%E5%91%8A/" style="font-size: 12.5px;">报告</a> <a href="/tags/%E6%8E%92%E7%89%88/" style="font-size: 13.75px;">排版</a> <a href="/tags/%E6%95%85%E4%B9%A1/" style="font-size: 10.63px;">故乡</a> <a href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 10.63px;">数据库</a> <a href="/tags/%E6%A0%87%E6%9C%AC%E9%A6%86/" style="font-size: 14.38px;">标本馆</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%88%86%E7%B1%BB/" style="font-size: 10px;">植物分类</a> <a href="/tags/%E6%A4%8D%E7%89%A9%E5%AD%A6%E5%90%8D/" style="font-size: 10px;">植物学名</a> <a href="/tags/%E6%B1%82%E8%81%8C/" style="font-size: 10px;">求职</a> <a href="/tags/%E7%89%88%E6%9C%AC%E6%8E%A7%E5%88%B6/" style="font-size: 11.25px;">版本控制</a> <a href="/tags/%E7%94%9F%E6%80%81/" style="font-size: 16.88px;">生态</a> <a href="/tags/%E7%94%9F%E6%80%81%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">生态资讯</a> <a href="/tags/%E7%A0%94%E7%A9%B6%E8%B5%84%E8%AE%AF/" style="font-size: 10px;">研究资讯</a> <a href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2/" style="font-size: 10px;">系统发育</a> <a href="/tags/%E7%BB%9F%E8%AE%A1/" style="font-size: 18.13px;">统计</a> <a href="/tags/%E7%BC%96%E7%A8%8B/" style="font-size: 10.63px;">编程</a> <a href="/tags/%E8%A1%8C%E6%91%84%E5%8C%86%E5%8C%86/" style="font-size: 15px;">行摄匆匆</a> <a href="/tags/%E8%A7%81%E9%97%BB%E5%BD%95/" style="font-size: 10px;">见闻录</a> <a href="/tags/%E8%AF%97%E6%AD%8C/" style="font-size: 15.63px;">诗歌</a> <a href="/tags/%E8%BD%AF%E4%BB%B6/" style="font-size: 19.38px;">软件</a> <a href="/tags/%E8%BF%9B%E5%8C%96/" style="font-size: 10.63px;">进化</a> <a href="/tags/%E9%9A%8F%E6%83%B3/" style="font-size: 11.25px;">随想</a> <a href="/tags/%E9%9A%8F%E7%AC%94/" style="font-size: 20px;">随笔</a> <a href="/tags/%E9%A9%B3%E8%AE%BA/" style="font-size: 10px;">驳论</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Archives</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/03/">March 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2021/02/">February 2021</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/12/">December 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/11/">November 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/06/">June 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/05/">May 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/02/">February 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">January 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/10/">October 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/09/">September 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/07/">July 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/06/">June 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/05/">May 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/04/">April 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/03/">March 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/02/">February 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/01/">January 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/12/">December 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/11/">November 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/10/">October 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/09/">September 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/08/">August 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/07/">July 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/06/">June 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/05/">May 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/04/">April 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/03/">March 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/02/">February 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/01/">January 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/11/">November 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/10/">October 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/09/">September 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/08/">August 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/06/">June 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/05/">May 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/04/">April 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/02/">February 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/12/">December 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/11/">November 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/09/">September 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/06/">June 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/05/">May 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/04/">April 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/03/">March 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/01/">January 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/11/">November 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/10/">October 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/09/">September 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/07/">July 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/05/">May 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/04/">April 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/03/">March 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/02/">February 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/01/">January 2015</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/12/">December 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/11/">November 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/10/">October 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/09/">September 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/07/">July 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/05/">May 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/03/">March 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/02/">February 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2014/01/">January 2014</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/12/">December 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/10/">October 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/05/">May 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/04/">April 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/03/">March 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/02/">February 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2013/01/">January 2013</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/12/">December 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/10/">October 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/08/">August 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/07/">July 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/06/">June 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/05/">May 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/04/">April 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/02/">February 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2012/01/">January 2012</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/12/">December 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/11/">November 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/10/">October 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/08/">August 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/07/">July 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/05/">May 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/04/">April 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/03/">March 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2011/02/">February 2011</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/12/">December 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/11/">November 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/09/">September 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/08/">August 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/07/">July 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/06/">June 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/05/">May 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/04/">April 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/03/">March 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2010/01/">January 2010</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/08/">August 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/07/">July 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/06/">June 2009</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2009/05/">May 2009</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">Recent Posts</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2021/03/26/2021-3-26/">使用hexo生成和部署网站</a>
          </li>
        
          <li>
            <a href="/2021/03/25/2021-03-25/">用pkgdown为R程序包建立网站</a>
          </li>
        
          <li>
            <a href="/2021/03/18/2021-03-18/">为什么《xxxx在XX研究中取得进展》这类文章很少有人转发？</a>
          </li>
        
          <li>
            <a href="/2021/02/16/2021-02-16/">如何恢复森林？– 从嘉道理农场的森林恢复样地说起</a>
          </li>
        
          <li>
            <a href="/2020/12/30/2020-12-30/">群落系统发育研究不能脱离理论框架</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      
      &copy; 2021 Jinlong Zhang<br>
      Powered by <a href="https://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>

    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
    <a href="/about" class="mobile-nav-link">About</a>
  
</nav>
    


<script src="/js/jquery-3.4.1.min.js"></script>



  
<script src="/fancybox/jquery.fancybox.min.js"></script>




<script src="/js/script.js"></script>





  </div>
</body>
</html>